_posts/2018-06-25-refactoring-with-design-patterns.markdown

```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: "Refactoring: clean your ruby code with design patterns"
+date: 2018-06-25 12:00:00
+categories: ["code-refactor"]
+author: "cleiviane"
+---
+
+Code refactoring can be defined as <strong>*“the process of introducing small and incremental changes to leave the code in a better state than it was.”*</strong>. When refactoring your code you have to consider two things: none of new functionality should be added and the external behavior should not be affected.
+
+One of the Ruby on Rails developer challenges is to keep your code clean, simple and easy to maintain and that is why we are always refactoring our code.
+
+There are several techniques that a developer can follow to improve his code by code refactoring, such as extract method, move method, move field, switch statements, etc. If you are not familiarized with them, please visit the [Refactoring Guru site](https://refactoring.guru/).
+
+<!--more-->
+
+<h2 id="design-patterns">Design Patterns</h2>
+
+Another technique developers try to follow is to apply good design patterns to its code. A design pattern can be described as “typical solutions to commonly occurring problems in software design. They are blueprints, that can be taken and customized to solve a particular design problem in your code”. So design patterns are not rules but guides that can help you to find the best solution given a particular situation. That is why, as pragmatic developers, we have always a good design pattern on the hand.
+
+Explain each pattern in detail is out of this article’s scope, so if you are not familiar with the presented design patterns, a good reference is the book Design Patterns: Elements of Reusable Object-Oriented
+Software, written by the famous Gang of Four (GoF).
+
+Since we have cataloged over 23 design patterns, our intention here is to show our readers the most used and how they can be useful to clean our ruby code.
```

Comment:
`Since we have cataloged over 23 design patterns` sounds weird, we didn't catalogue them. Should be `Since GoF catalogued around 23 design patterns`, or something similar.